#!/usr/bin/env python3
"""
gFlow æ’ä»¶ç¤ºä¾‹ - Echo Plugin

è¿™æ˜¯ä¸€ä¸ªç®€å•çš„ gRPC æ’ä»¶ç¤ºä¾‹ï¼Œç”¨äºæµ‹è¯• gFlow çš„æ’ä»¶ç³»ç»Ÿã€‚
å®ƒæ¥æ”¶è¾“å…¥å¹¶å°†å…¶å›æ˜¾ã€‚

ä½¿ç”¨æ–¹æ³•:
1. å®‰è£…ä¾èµ–: pip install grpcio grpcio-tools
2. ç”Ÿæˆ proto ä»£ç : python -m grpc_tools.protoc -I../../proto --python_out=. --grpc_python_out=. ../../proto/node_plugin.proto
3. è¿è¡Œ: python echo_plugin.py
"""

import grpc
from concurrent import futures
import time
import json
import sys
import os

# æ·»åŠ å½“å‰ç›®å½•åˆ°è·¯å¾„
sys.path.insert(0, os.path.dirname(__file__))

try:
    import node_plugin_pb2 as pb
    import node_plugin_pb2_grpc as pb_grpc
except ImportError:
    print("è¯·å…ˆç”Ÿæˆ proto ä»£ç :")
    print("python -m grpc_tools.protoc -I../../proto --python_out=. --grpc_python_out=. ../../proto/node_plugin.proto")
    sys.exit(1)


def value_to_python(value):
    """å°† proto Value è½¬æ¢ä¸º Python å€¼"""
    if value is None:
        return None
    
    kind = value.WhichOneof('kind')
    if kind == 'null_value':
        return None
    elif kind == 'string_value':
        return value.string_value
    elif kind == 'int_value':
        return value.int_value
    elif kind == 'double_value':
        return value.double_value
    elif kind == 'bool_value':
        return value.bool_value
    elif kind == 'bytes_value':
        return value.bytes_value
    elif kind == 'list_value':
        return [value_to_python(v) for v in value.list_value.values]
    elif kind == 'map_value':
        return {k: value_to_python(v) for k, v in value.map_value.fields.items()}
    return None


def python_to_value(val):
    """å°† Python å€¼è½¬æ¢ä¸º proto Value"""
    if val is None:
        return pb.Value(null_value=pb.NULL_VALUE)
    elif isinstance(val, str):
        return pb.Value(string_value=val)
    elif isinstance(val, bool):
        return pb.Value(bool_value=val)
    elif isinstance(val, int):
        return pb.Value(int_value=val)
    elif isinstance(val, float):
        return pb.Value(double_value=val)
    elif isinstance(val, bytes):
        return pb.Value(bytes_value=val)
    elif isinstance(val, list):
        return pb.Value(list_value=pb.ListValue(values=[python_to_value(v) for v in val]))
    elif isinstance(val, dict):
        return pb.Value(map_value=pb.MapValue(fields={k: python_to_value(v) for k, v in val.items()}))
    else:
        return pb.Value(string_value=str(val))


class EchoPluginService(pb_grpc.NodePluginServiceServicer):
    """Echo æ’ä»¶æœåŠ¡å®ç°"""
    
    def GetMetadata(self, request, context):
        """è¿”å›æ’ä»¶å…ƒæ•°æ®"""
        return pb.GetMetadataResponse(
            name="echo_plugin",
            display_name="Echo Plugin",
            description="ä¸€ä¸ªç®€å•çš„å›æ˜¾æ’ä»¶ï¼Œç”¨äºæµ‹è¯•",
            version="1.0.0",
            icon="MessageCircle",
            category=pb.CATEGORY_ACTION,
            node_type=pb.NODE_TYPE_PROCESSOR,
            input_parameters=[
                pb.ParameterDef(
                    name="message",
                    display_name="æ¶ˆæ¯",
                    type=pb.PARAM_TYPE_STRING,
                    description="è¦å›æ˜¾çš„æ¶ˆæ¯",
                    required=True,
                    default_value=pb.Value(string_value="Hello, World!"),
                    ui_type=pb.UI_TYPE_TEXTAREA,
                ),
                pb.ParameterDef(
                    name="prefix",
                    display_name="å‰ç¼€",
                    type=pb.PARAM_TYPE_STRING,
                    description="æ·»åŠ åˆ°æ¶ˆæ¯å‰çš„å‰ç¼€",
                    required=False,
                    default_value=pb.Value(string_value="[Echo]"),
                    ui_type=pb.UI_TYPE_TEXT,
                ),
                pb.ParameterDef(
                    name="delay",
                    display_name="å»¶è¿Ÿ(ç§’)",
                    type=pb.PARAM_TYPE_INT,
                    description="å¤„ç†å»¶è¿Ÿæ—¶é—´",
                    required=False,
                    default_value=pb.Value(int_value=0),
                    ui_type=pb.UI_TYPE_NUMBER,
                ),
            ],
            output_parameters=[
                pb.ParameterDef(
                    name="result",
                    display_name="ç»“æœ",
                    type=pb.PARAM_TYPE_STRING,
                    description="å›æ˜¾çš„æ¶ˆæ¯",
                ),
                pb.ParameterDef(
                    name="timestamp",
                    display_name="æ—¶é—´æˆ³",
                    type=pb.PARAM_TYPE_STRING,
                    description="å¤„ç†æ—¶é—´æˆ³",
                ),
            ],
            capabilities=pb.PluginCapabilities(
                supports_streaming=True,
                supports_cancel=False,
                supports_retry=True,
                requires_credential=False,
                max_concurrent=10,
                default_timeout_ms=30000,
            ),
        )
    
    def Init(self, request, context):
        """åˆå§‹åŒ–æ’ä»¶"""
        print(f"[Init] Node: {request.node_config.name if request.node_config else 'unknown'}")
        print(f"[Init] Workflow: {request.workflow_config.name if request.workflow_config else 'unknown'}")
        return pb.InitResponse(success=True)
    
    def Run(self, request, context):
        """æ‰§è¡Œæ’ä»¶é€»è¾‘"""
        # è§£æå‚æ•°
        params = {k: value_to_python(v) for k, v in request.parameters.items()}
        message = params.get('message', 'No message provided')
        prefix = params.get('prefix', '[Echo]')
        delay = params.get('delay', 0)
        
        print(f"[Run] Received message: {message}")
        
        # å‘é€æ—¥å¿—
        yield pb.RunResponse(
            type=pb.RESPONSE_TYPE_LOG,
            timestamp_ms=int(time.time() * 1000),
            log=pb.LogPayload(
                level=pb.LOG_LEVEL_INFO,
                message=f"å¼€å§‹å¤„ç†æ¶ˆæ¯: {message[:50]}...",
            ),
        )
        
        # æ¨¡æ‹Ÿå¤„ç†å»¶è¿Ÿ
        if delay > 0:
            for i in range(delay):
                yield pb.RunResponse(
                    type=pb.RESPONSE_TYPE_PROGRESS,
                    timestamp_ms=int(time.time() * 1000),
                    progress=pb.ProgressPayload(
                        current=i + 1,
                        total=delay,
                        percentage=((i + 1) / delay) * 100,
                        message=f"å¤„ç†ä¸­... {i + 1}/{delay}",
                    ),
                )
                time.sleep(1)
        
        # ç”Ÿæˆç»“æœ
        result = f"{prefix} {message}"
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        
        yield pb.RunResponse(
            type=pb.RESPONSE_TYPE_LOG,
            timestamp_ms=int(time.time() * 1000),
            log=pb.LogPayload(
                level=pb.LOG_LEVEL_INFO,
                message="å¤„ç†å®Œæˆ",
            ),
        )
        
        # å‘é€ç»“æœ
        yield pb.RunResponse(
            type=pb.RESPONSE_TYPE_RESULT,
            timestamp_ms=int(time.time() * 1000),
            result=pb.ResultPayload(
                output={
                    "result": python_to_value(result),
                    "timestamp": python_to_value(timestamp),
                    "original_message": python_to_value(message),
                    "prefix": python_to_value(prefix),
                },
                branch_index=0,
                status=pb.EXECUTION_STATUS_SUCCESS,
                duration_ms=delay * 1000 if delay else 10,
            ),
        )
    
    def Stop(self, request, context):
        """åœæ­¢æ‰§è¡Œ"""
        print(f"[Stop] Received stop request: {request.reason}")
        return pb.StopResponse(
            success=True,
            status=pb.STOP_STATUS_STOPPED,
            message="å·²åœæ­¢",
        )
    
    def TestCredential(self, request, context):
        """æµ‹è¯•å‡­è¯"""
        return pb.TestCredentialResponse(
            success=True,
            info={"message": "æ­¤æ’ä»¶ä¸éœ€è¦å‡­è¯"},
        )
    
    def HealthCheck(self, request, context):
        """å¥åº·æ£€æŸ¥"""
        return pb.HealthCheckResponse(
            status=pb.HEALTH_STATUS_HEALTHY,
            message="OK",
            plugin_version="1.0.0",
            protocol_version="2.0.0",
            supported_features=["streaming", "retry"],
        )


def serve(port=50051):
    """å¯åŠ¨ gRPC æœåŠ¡"""
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
        options=[
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.max_send_message_length', 50 * 1024 * 1024),     # 50MB
        ],
    )
    pb_grpc.add_NodePluginServiceServicer_to_server(EchoPluginService(), server)
    
    listen_address = f'[::]:{port}'
    server.add_insecure_port(listen_address)
    
    server.start()
    print(f"ğŸš€ Echo Plugin å¯åŠ¨æˆåŠŸ!")
    print(f"ğŸ“¡ ç›‘å¬åœ°å€: localhost:{port}")
    print(f"ğŸ“¦ æ’ä»¶ç±»å‹: echo_plugin")
    print("")
    print("åœ¨ gFlow çš„ config/plugins.yaml ä¸­æ·»åŠ ä»¥ä¸‹é…ç½®:")
    print(f"""
plugins:
  - name: "Echo Plugin"
    kind: "echo_plugin"
    endpoint: "localhost:{port}"
    enabled: true
    health_check: true
    description: "å›æ˜¾æ’ä»¶"
    category: "action"
    icon: "MessageCircle"
    color: "blue"
""")
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\nğŸ‘‹ æ­£åœ¨å…³é—­æœåŠ¡...")
        server.stop(0)


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser(description='Echo Plugin for gFlow')
    parser.add_argument('--port', type=int, default=50051, help='ç›‘å¬ç«¯å£ (é»˜è®¤: 50051)')
    args = parser.parse_args()
    
    serve(args.port)
